pkg/sql/sem/builtins/builtins.go

```diff
@@ -11296,3 +11345,55 @@ func bitmaskOp(aStr, bStr string, op func(byte, byte) byte) (*tree.DBitArray, er
 
 	return tree.ParseDBitArray(string(buf))
 }
+
+func makeTimestampStatementBuiltinOverload(withTZ bool, withInputTZ bool) tree.Overload {
+	// If we are not creating a timestamp with a timezone, we shouldn't expect an input timezone
+	if !withTZ {
+		withInputTZ = false
+	}
+	info := "Create timestamp"
+	typs := tree.ParamTypes{{Name: "year", Typ: types.Int}, {Name: "month", Typ: types.Int}, {Name: "day", Typ: types.Int},
+		{Name: "hour", Typ: types.Int}, {Name: "min", Typ: types.Int}, {Name: "sec", Typ: types.Float}}
+	returnTyp := types.Timestamp
+	if withTZ {
+		info += "with time zone from year, month, day, hour, minute and seconds fields (negative years signify BC). If " +
+			"timezone is not specified, the current time zone is used."
+		returnTyp = types.TimestampTZ
+		if withInputTZ {
+			typs = append(typs, tree.ParamType{Name: "timezone", Typ: types.String})
+		}
+	} else {
+		info += "from year, month, day, hour, minute, and seconds fields (negative years signify BC)."
+	}
+	return tree.Overload{
+		Types:      typs,
+		ReturnType: tree.FixedReturnType(returnTyp),
+		Fn: func(_ context.Context, evalCtx *eval.Context, args tree.Datums) (tree.Datum, error) {
+			year := int(tree.MustBeDInt(args[0]))
+			month := time.Month(int(tree.MustBeDInt(args[1])))
+			day := int(tree.MustBeDInt(args[2]))
+			location := evalCtx.GetLocation()
+			var err error
+			if withInputTZ && withTZ {
+				location, err = timeutil.TimeZoneStringToLocation(string(tree.MustBeDString(args[6])), timeutil.TimeZoneStringToLocationPOSIXStandard)
+				if err != nil {
+					return nil, err
+				}
+			}
+			if year == 0 {
+				return nil, pgerror.New(pgcode.InvalidParameterValue, "year value of 0 is not valid")
+			}
+			hour := int(tree.MustBeDInt(args[3]))
+			min := int(tree.MustBeDInt(args[4]))
+			sec := float64(tree.MustBeDFloat(args[5]))
+			truncatedSec := math.Floor(sec)
+			nsec := math.Mod(sec, truncatedSec) * float64(time.Second)
+			if withTZ {
+				return tree.MakeDTimestampTZ(time.Date(year, month, day, hour, min, int(truncatedSec), int(nsec), location), time.Microsecond)
+			}
+			return tree.MakeDTimestamp(time.Date(year, month, day, hour, min, int(truncatedSec), int(nsec), location), time.Microsecond)
+		},
+		Info:       info,
+		Volatility: volatility.Stable,
```

Comment:
you can use a query like this in PG to see what the volatility should be:
```
select pg_get_function_arg_default(p.oid, 1), proname, pronamespace, n.nspname, proargtypes, prorettype, provolatile, proleakproof from pg_proc p join pg_namespace n on p.pronamespace = n.oid where proname ilike 'make_%';
```

seems like only `make_timestamptz` is stable, and the other `make_*` functions are immutable.

`date_trunc(.., timestamptz)` is stable, while the other versions are immutable.

(Check out the comment on the Volatility type if you want to see what this means)